cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java

```diff
@@ -1010,21 +1010,31 @@ private QueryBuilder createMainQuery(SearchRequest request) {
     // all terms must occur in the text field as selected by the scope in the search request.
     String textField = request.getScope() == null ? TEXT_FIELD : request.getScope().name().toLowerCase();
 
-    // split the query into its terms and iterate over all terms
-    Iterable<String> terms = Splitter.on(SPACE_SEPARATOR_PATTERN)
-      .omitEmptyStrings().trimResults().split(request.getQuery());
-    List<QueryBuilder> termQueries = new ArrayList<>();
-    for (String term : terms) {
-      termQueries.add(createTermQuery(term, textField, request));
+    // split the query into its parsed terms and create corresponding QueryBuilders for each
+    Map<QueryBuilder, QueryTerm> termQueries = new HashMap<>();
```

Comment:
my concern here how QueryTerm implements hashCode and equals, I doubt that it was designed to be used in a hashmap. Since all you do in this map is add values and iterate, could you use an ArrayList<ImmutablePair<QueryBuilder, QueryTerm>> instead?

Reply:
Addressed in the latest commit.